### PR TITLE
cilium, bpf: fix panic when run with newer LLVM

### PR DESCRIPTION
**Pull request by borkmann**
```
While debugging another issue I ran into the following panic wrt
ELF templating:

  # daemon/cilium-agent --kvstore consul --kvstore-opt consul.address=127.0.0.1:8500 --enable-ipv6=false --masquerade=true --auto-direct-node-routes=true --disable-envoy-version-check
  [...]
  panic: runtime error: index out of range

  goroutine 318 [running]:
  github.com/cilium/cilium/pkg/elf.(*symbols).extractFrom(0xc0006520d0, 0xc000f28320, 0xc000f28320, 0x0)
  	/root/go/src/github.com/cilium/cilium/pkg/elf/symbols.go:170 +0x1357
  github.com/cilium/cilium/pkg/elf.NewELF(0x317a580, 0xc00000f518, 0xc0001090a0, 0x0, 0x0, 0xc0001090a0)
  	/root/go/src/github.com/cilium/cilium/pkg/elf/elf.go:77 +0x145
  github.com/cilium/cilium/pkg/elf.Open(0xc000afd5c0, 0x52, 0xc0000a6c00, 0x7f42126307d8, 0xc0006f20f0)
  	/root/go/src/github.com/cilium/cilium/pkg/elf/elf.go:97 +0x1eb
  github.com/cilium/cilium/pkg/datapath/loader.CompileOrLoad(0x31b6f80, 0xc0000a6c00, 0x31db1c0, 0xc0006f20f0, 0xc000fdf7c0, 0x0, 0x0)
  	/root/go/src/github.com/cilium/cilium/pkg/datapath/loader/loader.go:123 +0xd6
  github.com/cilium/cilium/pkg/endpoint.(*Endpoint).realizeBPFState(0xc000d46000, 0xc000fdf680, 0xc000fdf888, 0xc000fdf680, 0xc0012d5ad8)
  	/root/go/src/github.com/cilium/cilium/pkg/endpoint/bpf.go:488 +0x30d
  github.com/cilium/cilium/pkg/endpoint.(*Endpoint).regenerateBPF(0xc000d46000, 0x31d7540, 0xc0007b21c0, 0xc000fdf680, 0x0, 0x0, 0x0, 0x0)
  	/root/go/src/github.com/cilium/cilium/pkg/endpoint/bpf.go:415 +0x273
  github.com/cilium/cilium/pkg/endpoint.(*Endpoint).regenerate(0xc000d46000, 0x31d7540, 0xc0007b21c0, 0xc000fdf680, 0x0, 0x0)
  	/root/go/src/github.com/cilium/cilium/pkg/endpoint/policy.go:323 +0x704
  github.com/cilium/cilium/pkg/endpoint.(*EndpointRegenerationEvent).Handle(0xc00139d8c0, 0xc0000a6ba0)
  	/root/go/src/github.com/cilium/cilium/pkg/endpoint/events.go:54 +0x21b
  github.com/cilium/cilium/pkg/eventqueue.(*EventQueue).Run.func1()
  	/root/go/src/github.com/cilium/cilium/pkg/eventqueue/eventqueue.go:236 +0x144
  sync.(*Once).Do(0xc0004f0e78, 0xc0010df540)
  	/usr/local/go/src/sync/once.go:44 +0xb3
  created by github.com/cilium/cilium/pkg/eventqueue.(*EventQueue).Run
  	/root/go/src/github.com/cilium/cilium/pkg/eventqueue/eventqueue.go:225 +0xa9
  level=fatal msg="Agent pipe unexpectedly closed, shutting down" subsys=cilium-node-monitor

Reason is that newer LLVM changed BPF backend to emit symbols under
OBJECT instead of NOTYPE symbol scope. Also, BTF line info is not
recognized by golang's default ELF parser, so accessing section for
it will lead to the oob access panic (sym.Section is 64k here). Just
skip these.

  # llc --version
  LLVM (http://llvm.org/):
  LLVM version 9.0.0svn
  Optimized build.
  Default target: x86_64-unknown-linux-gnu
  Host CPU: haswell

  Registered Targets:
    bpf    - BPF (host endian)
    bpfeb  - BPF (big endian)
    bpfel  - BPF (little endian)
    x86    - 32-bit X86: Pentium-Pro and above
    x86-64 - 64-bit X86: EM64T and AMD64

Symtab from bpf_netdev:

  # readelf -a /var/run/cilium/state/bpf_netdev.o
  [...]
    30: 00000000000000a8    28 OBJECT  GLOBAL DEFAULT   14 cilium_calls_netdev_2
    31: 0000000000000000    28 OBJECT  GLOBAL DEFAULT   14 cilium_events
    32: 00000000000000c4    28 OBJECT  GLOBAL DEFAULT   14 cilium_ipcache
    33: 000000000000001c    28 OBJECT  GLOBAL DEFAULT   14 cilium_lxc
    34: 0000000000000038    28 OBJECT  GLOBAL DEFAULT   14 cilium_metrics
    35: 0000000000000054    28 OBJECT  GLOBAL DEFAULT   14 cilium_policy
    36: 0000000000000070    28 OBJECT  GLOBAL DEFAULT   14 cilium_policy_reserved_2
    37: 000000000000008c    28 OBJECT  GLOBAL DEFAULT   14 cilium_proxy4
  [...]

Symtab from bpf_lxc:

  # readelf -a /var/run/cilium/state/templates/c4406a7451ccb067746c874b71aaf19266ce7212/bpf_lxc.o
  [...]
   285: 000000000000001c     4 OBJECT  GLOBAL DEFAULT   11 LXC_ID
   286: 0000000000000010     4 OBJECT  GLOBAL DEFAULT   11 LXC_IPV4
   287: 0000000000000000     4 OBJECT  GLOBAL DEFAULT   11 LXC_IP_1
   288: 0000000000000004     4 OBJECT  GLOBAL DEFAULT   11 LXC_IP_2
   289: 0000000000000008     4 OBJECT  GLOBAL DEFAULT   11 LXC_IP_3
   290: 000000000000000c     4 OBJECT  GLOBAL DEFAULT   11 LXC_IP_4
   291: 0000000000000014     4 OBJECT  GLOBAL DEFAULT   11 NODE_MAC_1
   292: 0000000000000018     4 OBJECT  GLOBAL DEFAULT   11 NODE_MAC_2
   293: 0000000000000020     4 OBJECT  GLOBAL DEFAULT   11 SECLABEL
   294: 0000000000000024     4 OBJECT  GLOBAL DEFAULT   11 SECLABEL_NB
   295: 0000000000000000     4 OBJECT  GLOBAL DEFAULT   13 ____license
  [...]

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

```

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/7648)

### CODE REVIEW

**Comment by borkmann** [comment]
test-me-please

**Comment by borkmann** [comment]
Hit #7282.

**Comment by borkmann** [comment]
test-me-please

File: pkg/elf/symbols.go [hunk 1]
```diff
@@ -196,6 +194,13 @@ func (s *symbols) extractFrom(e *elf.File) error {
 			symOffset := strtab.Offset + symOffsetInStrtab
 			stringOffsets[sym.Name] = newString(sym.Name, symOffset)
 			log.WithField(fieldSymbol, sym.Name).Debugf("Found symbol with offset %d", symOffset)
+		case !isGlobalData(sym):
+			// LBB is a common llvm symbol prefix (basic block);
+			// Don't flood the logs with messages about it.
+			if !strings.HasPrefix(sym.Name, "LBB") {
+				log.Debugf("Skipping %s", sym.Name)
+			}
+			continue
```
Comment (joestringer) [inline pkg/elf/symbols.go]: Is it meaningful to shift this global section check down here?

**Review by joestringer — APPROVED** [review]

File: pkg/elf/symbols.go (hunk 1 above)
Comment (borkmann) [reply pkg/elf/symbols.go]: Thinking about it again, nope, removed.

**Comment by borkmann** [comment]
test-me-please

**Pull request merged**